File: app/src/main/java/com/salman/bitclock/ui/timer/TimerViewModel.java
package com.salman.bitclock.ui.timer;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.salman.bitclock.data.models.Timer;
import com.salman.bitclock.data.repository.TimerRepository;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class TimerViewModel extends ViewModel {

    private final TimerRepository repository;
    private final LiveData<List<Timer>> allTimers;
    private final ExecutorService executorService = Executors.newSingleThreadExecutor();

    @Inject
    public TimerViewModel(TimerRepository repository) {
        this.repository = repository;
        this.allTimers = repository.getAllTimers();
    }

    public LiveData<List<Timer>> getAllTimers() {
        return allTimers;
    }

    public void insert(Timer timer) {
        executorService.execute(() -> repository.insert(timer));
    }

    public void update(Timer timer) {
        executorService.execute(() -> repository.update(timer));
    }

    public void delete(Timer timer) {
        executorService.execute(() -> repository.delete(timer));
    }
}

File: app/src/main/kotlin/com/salman/bitclock/ui/timer/TimerViewModel.kt
package com.salman.bitclock.ui.timer

import android.content.Context
import android.content.Intent
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.salman.bitclock.data.models.Timer
import com.salman.bitclock.data.repository.TimerRepository
import com.salman.bitclock.services.TimerService
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class TimerViewModel @Inject constructor(
    private val repository: TimerRepository,
    @ApplicationContext private val context: Context
) : ViewModel() {

    val timers: StateFlow<List<Timer>> = repository.getAllTimers()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    fun addTimer(name: String, durationMs: Long) {
        viewModelScope.launch {
            val timer = Timer(
                name = if (name.isEmpty()) "Timer" else name,
                initialDurationMs = durationMs,
                remainingMs = durationMs
            )
            repository.insert(timer)
        }
    }

    fun startTimer(timerId: Int) {
        val intent = Intent(context, TimerService::class.java).apply {
            action = TimerService.ACTION_START
            putExtra(TimerService.EXTRA_TIMER_ID, timerId)
        }
        context.startService(intent)
    }

    fun pauseTimer(timerId: Int) {
        val intent = Intent(context, TimerService::class.java).apply {
            action = TimerService.ACTION_PAUSE
            putExtra(TimerService.EXTRA_TIMER_ID, timerId)
        }
        context.startService(intent)
    }

    fun resetTimer(timerId: Int) {
        val intent = Intent(context, TimerService::class.java).apply {
            action = TimerService.ACTION_RESET
            putExtra(TimerService.EXTRA_TIMER_ID, timerId)
        }
        context.startService(intent)
    }

    fun deleteTimer(timer: Timer) {
        viewModelScope.launch {
            repository.delete(timer)
            // Also ensure it's stopped in service
            val intent = Intent(context, TimerService::class.java).apply {
                action = TimerService.ACTION_STOP
                putExtra(TimerService.EXTRA_TIMER_ID, timer.id)
            }
            context.startService(intent)
        }
    }
}
